Extract execution engine

lptr opened the pull request:
This is an attempt to extract an execution engine independent of tasks.

Roles of the execution engine:

* it is responsible to make sure a _unit of work_ is _realized,_ meaning it has either been executed, or some shortcut mechanism produced the same results as if it was executed
* it is compatible with tasks _and_ artifact transforms (and potentially other work like component metadata rules etc.)
* it handles incremental execution, i.e. can invoke work with a list of changes when requested
* it handles incremental build, i.e. keeps track of the work's history, and marks it as `UP-TO-DATE` when appropriate
* it handles offloading work to the build cache when possible

What it isn't:

* it is _not_ responsible for _scheduling_ the work, deciding on dependencies etc.
* execution is synchronous, no attempt is made in the executor to parallelize any of the work, or any handling of locks to assure that the work can happen uninterrupted

It's probably best to review the PR commit-by-commit.

wolfs commented on subprojects/core/src/test/groovy/org/gradle/cache/internal/DefaultFileContentCacheFactoryTest.groovy:
```diff
@@ -16,14 +16,14 @@
 
-import org.gradle.api.internal.changedetection.state.InMemoryCacheDecoratorFactory
-import org.gradle.api.internal.tasks.execution.TaskOutputChangesListener
+
```
Nit: extra new line.

wolfs commented on subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/AbstractFingerprintChanges.java:
```diff
@@ -14,38 +14,27 @@
-    protected final TaskExecution previous;
-    protected final TaskExecution current;
+public abstract class AbstractFingerprintChanges implements ChangeContainer {
+    protected final ImmutableSortedMap<String, ? extends FileCollectionFingerprint> previous;
```
The `? extends` is not necessary anymore for `previous` and `current`.

lptr replied:
Good point, fixed.

wolfs commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ResolveTaskOutputCachingStateExecuter.java:
```diff
@@ -16,36 +16,168 @@
+
+        if (overlappingOutputs != null) {
+            return DefaultTaskOutputCachingState.disabled(TaskOutputCachingDisabledReasonCategory.OVERLAPPING_OUTPUTS,
+                String.format("Gradle does not know how file '%s' was created (output property '%s'). Task output caching requires exclusive access to output paths to guarantee correctness.",
```
It would be great to be able to retain a relative path here. At least in Gradle Enterprise this path should be shown as relative.

lptr replied:
I looked at the corresponding test and I don't see why a relative path makes things more understandable. The user only has the path of the task and a relative path, so if they want to take a look at the actual file they need to do some digging, An absolute path seems more useful to me.

wolfs replied:
Note that the relative path was relative to the project - so it is not hard to find it. An absolute path is very unwieldy, especially when displayed in build scans. I we use an absolute path here, then the build scan side should take care to convert it into a relative path to the root project, if only for deduplication reasons.

@big-guy suggesting using a relative path quite some time ago: https://code-review.gradle.org/cru/REVIEW-6515#CFR-109810

wolfs commented on subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/CachedPlatformScalaCompileIntegrationTest.groovy:
```diff
@@ -16,6 +16,7 @@
 
 package org.gradle.language.scala
 
+
```
Nit: Unexpected newline.

wolfs commented on subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/CachedPlatformScalaCompileIntegrationTest.groovy:
```diff
@@ -77,7 +78,7 @@ class CachedPlatformScalaCompileIntegrationTest extends AbstractCachedCompileInt
         compiledJavaClass.exists()
         compiledScalaClass.exists()
-        output.contains "Caching disabled for task ':compileMainJarMainScala': Gradle does not know how file 'build"
+        output.contains "Caching disabled for task ':compileMainJarMainScala': Gradle does not know how file '"
```
I think we should check the file name here, not only that there are overlapping outputs.

lptr replied:
Re-added check for the `build` directory as it was before.

wolfs commented on subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/ShortCircuitTaskArtifactStateRepository.java:
```diff
@@ -132,15 +127,20 @@ public void afterOutputsRemovedBeforeTask() {
-        public void snapshotAfterLoadedFromCache(ImmutableSortedMap<String, CurrentFileCollectionFingerprint> newOutputFingerprints, OriginMetadata originMetadata) {
-            delegate.snapshotAfterLoadedFromCache(newOutputFingerprints, originMetadata);
+        public OverlappingOutputs getOverlappingOutputs() {
+            return null;
```
I don't understand why we can avoid calculating overlapping outputs in `RerunTaskArtifactState`. We could just use `delegate.getOverlappingOutputs()`.

It seems a little bit strange returning null here. Can we move this state maybe somewhere else?

lptr replied:
This is an error. I wonder how no test caught this.

wolfs commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java:
```diff
@@ -42,47 +46,53 @@
-                        if (!cleanupDirectories && file.isDirectory()) {
-                            continue;
+                for (FileCollectionFingerprint outputFingerprints : outputFiles.values()) {
+                    for (String outputRootPath : outputFingerprints.getFingerprints().keySet()) {
```
Why is `root` part of the name `outputRootPath`? This seems to be an `outputPath` or `outputFilePath` anywhere in the output.

lptr replied:
You're right, this is a leftover. Removed.

wolfs commented on subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/SkipUpToDateTaskExecuterTest.groovy:
```diff
@@ -16,48 +16,48 @@
 
 package org.gradle.api.internal.tasks.execution
 
+
```
Nit: Unnecessary newline.

wolfs commented on subprojects/execution/src/main/java/org/gradle/internal/execution/impl/steps/ExecuteStep.java:
```diff
@@ -0,0 +1,85 @@
+            failure = new ExecutionException(work, t);
+        }
+
+        boolean interrupted = Thread.interrupted();
```
Why do we need the interrupt handling here? There should be at least a comment what we are doing here.

lptr replied:
OK, added a comment. I want to discuss this with @oehme as well.

oehme replied:
That's why I originally extracted the snapshot step and put it above the interrupt handling step, so we wouldn't have to do workarounds like this.

lptr replied:
Makes sense, thanks, I'll change it before merging.

lptr replied:
@oehme can you please take another look?

lptr replied:
OK, now you can look again. :) Interrupts are hard! I still don't like how we make the build fail when timeout is up, though...

wolfs commented on subprojects/build-cache/src/main/java/org/gradle/caching/internal/CacheableEntity.java:
```diff
@@ -16,9 +16,11 @@
  */
-public interface CacheableEntity {
+public interface CacheableEntity extends Describable {
     String getPath();
```
Shall we call this `identity` now?

lptr replied:
Renamed, along with the property in the origin metadata file.

wolfs commented on subprojects/execution/src/main/java/org/gradle/internal/execution/impl/steps/CacheStep.java:
```diff
@@ -0,0 +1,192 @@
+    public org.gradle.internal.execution.ExecutionResult execute(C context) {
+        return context.getCacheHandler()
+            .load(cacheKey -> load(context.getWork(), cacheKey))
+            .map(loadResult -> (ExecutionResult) new ExecutionResult() {
```
I think this class would deserve an actual type, something like `LoadedFromCache`. Then this line would simplify to `.map(LoadedFromCache::new)`, right?

lptr replied:
I think you'd still need the cast because of the `orElseGet()`. Let's do this as part of reviewing the codebase for Java 8 compliance later.

wolfs commented on subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java:
```diff
@@ -0,0 +1,57 @@
+    CacheHandler createCacheHandler();
+
+    interface OutputVisitor {
+        void visitOutput(String name, OutputType type, FileCollection files, @Nullable File root);
```
The signature is now confusing for me. IIUC, `files` are the root files out the outputs, while `root` is the single root output if there is only one root for the output, right?
Can we make the names clearer?

lptr replied:
Good point, we don't need this complexity here.

lptr commented:
@wolfs I addressed all your comments except for the relative path when reporting overlapping outputs. If you are okay with resolving that separately, I'd go and merge this now.

wolfs commented on subprojects/build-cache/src/main/java/org/gradle/caching/internal/origin/OriginMetadataFactory.java:
```diff
@@ -76,7 +75,7 @@ public void execute(OutputStream outputStream) {
                 Properties properties = new Properties();
                 properties.setProperty(BUILD_INVOCATION_ID_KEY, currentBuildInvocationId.asString());
                 properties.setProperty(TYPE_KEY, entry.getClass().getCanonicalName());
```
Why don't we use `getName()` instead of `getCanonicalName()`?

lptr replied:
No idea, but it makes sense to have a top-level class implement the executable thingies, so that we can look it up easily.

wolfs replied:
+1 to the top level class. So your argument is that it doesn't make a difference if we use `getName` or `getCanonicalName`?

lptr replied:
Basically, they return the same, except `getCannonicalName()` returns `null` for non-top-level classes, which makes this code fail, which is a plus, so I'd leave it like this.

wolfs replied:
I don't see why this code would fail - it would write `null` into the property and store an empty property in the `METADATA` properties file.
I think we should either make it fail or use `getName()`.

wolfs commented on subprojects/base-services/src/main/java/org/gradle/internal/MutableLong.java:
```diff
@@ -0,0 +1,52 @@
+        this.value = value;
+    }
+
+    public long incrementAndGet(long delta) {
```
I think an `increment` method should be enough. Do you think the `incrementAndGet` method makes sense, given that we don't need any concurrency guarantees?

lptr replied:
I wanted it as a drop-in replacement for the typical use of `Atomic*`. Do you feel strongly about this?

lptr replied:
Ah, just realized the atomic API is different, changing now.

wolfs commented on subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPacker.java:
```diff
@@ -203,7 +201,7 @@ public String apply(@Nullable CacheableTree tree) {
                 boolean missing = matcher.group(1) != null;
                 String childPath = matcher.group(3);
-                tarEntry = unpackTree(tree, tarInput, tarEntry, childPath, missing, snapshots, entries);
+                tarEntry = unpackTree(treeName, tree.getType(), tree.getRoot(), tarInput, tarEntry, childPath, missing, snapshots, entries);
```
Should we use a `MutableLong` for entries here instead of an `AtomicInteger`?

lptr replied:
Good point. Why did we use an integer in any case?
